### DISCUSSION TITLE
Seeking migration path towards graphql-java 20.0 scalar coercion

### DISCUSSION THREAD

**Discussion by Ernir** [graphql-java/graphql-java#3057] [Q&A] [answered]
I am using a few simple custom scalars in an application. As [documented](https://www.graphql-java.com/documentation/scalars#writing-your-own-custom-scalars), this is done by providing a class that implements the [Coercing](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/schema/Coercing.java) interface.

For reference, this is one of them, defining a scalar representing a `java.time.Instant`:
```java
public final class InstantScalar implements Coercing<Instant, String> {
  @Override
  public String serialize(Object dataFetcherResult) {
    if (dataFetcherResult instanceof Instant) {
      return dataFetcherResult.toString();
    } else {
      throw new CoercingSerializeException(dataFetcherResult + " is not a valid Instant");
    }
  }

  @Override
  public Instant parseValue(Object input) {
    try {
      return Instant.parse(input.toString());
    } catch (RuntimeException exception) {
      throw new CoercingParseValueException(
          input + " is not a valid ISO date time value.", exception);
    }
  }

  @Override
  public Instant parseLiteral(Object input) {
    if (input instanceof StringValue stringValue) {
      try {
        return parseValue(stringValue.getValue());
      } catch (RuntimeException exception) {
        throw new CoercingParseLiteralException(
            input + " is not a valid ISO date time literal.", exception);
      }
    }
    throw new CoercingParseLiteralException(
        input + " must be an instance of graphql.language.StringValue.");
  }
}
```

In the [graphql-java 20.0 release notes](https://github.com/graphql-java/graphql-java/releases/tag/v20.0), it is described that there are breaking changes to coercion, but I am unsure how to adapt to those changes.

The example from the documentation and my class there above now both `@Override` methods of the `Coercing` interface which have been marked as `@Deprecated`.

How do I provide custom scalar coercion in graphql-java 20.0 without depending on the deprecated APIs?

**Comment by dondonz** [comment, marked as answer]
Hi @Ernir,

There were two recent changes to scalars:
1. Coercing interface change to enable internationalization (i18n) of error messages, released in v20.0 https://github.com/graphql-java/graphql-java/pull/2912
2. Breaking changes to built-in scalar `parseValue` coercion highlighted in the v20.0 release notes

The Coercing interface was changed to support i18n for coercing error messages, new methods take `Locale` and `GraphQLContext` as arguments. Have a look at built-in scalar Coercing implementations for ideas, for example Boolean: https://github.com/graphql-java/graphql-java/blob/v20.0/src/main/java/graphql/scalar/GraphqlBooleanCoercing.java

The breaking change highlighted at the top of the v20.0 release notes relates to the built-in scalars only, not for custom scalars. As more info, we made this change to align the built-in scalar `parseValue` coercion with the JS reference implementation, which made input types stricter.

You rightly point out our documentation is out of date! For anyone reading this, documentation PRs are always welcome on the `graphql-java-page` repo https://github.com/graphql-java/graphql-java-page

I hope that helps!

**Reply by ferdinand-swoboda** [reply]
@dondonz Thanks for your response and explanation. Indeed, the changes to the `Coercing` interface tripped us up.

Looking at the `GraphqlBooleanCoercing` example that you referenced that would still mean that custom scalars have to implement a method signature that is deprecated instead of, say, only implement the new method signature. 
Or am I missing something?

The reason we are bringing this up is because we have a policy of avoiding the use of deprecated APIs wherever possible and it seems this is not possible here.

**Reply by bbakerman** [reply]
It seems we have not `default` ed the deprecated methods which is an oversight on our behalf.  And hence any new scalar has to implement them even though wont be called direct.  The graphql-java code will always call the non deprecated method BUT it will typically call back to the old.

We should add `default` to the old deprecated methods so that new scalars do not have to implement them.  The fact is you still need to implement the new "default" methods BECAUSE for backwards compat reasons they have to be defaulted to call back to the old.

BUT of course the calling code ONLY calls the new and hence if you implement them, they will be non deprecated.

**Reply by bbakerman** [reply]
See https://github.com/graphql-java/graphql-java/pull/3063

**Reply by Ernir** [reply]
Thank you, this explains the situation fully! The confusing point was indeed that it seemed like there was no way to avoid overriding the deprecated methods.

Thanks for the PR, will follow progress there. :)
